[FEAT] Kakao Login 프로세스 구현

HyeonChae1104 opened the pull request:
## 주요 변경사항 ✔️

1. `FailurePage`: 로그인 실패 시 리다이렉트 되는 페이지
2. `SuccessPage` : 로그인 성공 시 리다이렉트 되는 페이지
3. `useSuccessLogin.ts` : 로그인 성공 시 백엔드에서 전달 받은 쿠키에서 accessToken을 확인하여 localStorage에 저장한 후 보안을 위해 삭제합니다.

4. `useHandleKakaoLogin` : 
- 로그인 실패 시 login page로 이동
- GET요청을 통해 백엔드로 리다이렉트

5. `useFetchWithToken`
연우님이 토큰 재발급에 대한 부분을 이렇게 처리해보면 어떨지 제안을 주셔서 회의 후 구현해보았습니다!
- 401 응답 시 토큰 재발급을 위한 useFetchWithToken 추가
- accessToken 만료 시 토큰 재발급을 처리하는 useFetchWithToken 훅 생성
- Axios 인터셉터를 통해 토큰 재발급 후 요청을 자동 재시도하도록 구현
- 토큰 재발급 실패 시 로그인 페이지로 리다이렉트


## 리뷰어에게... 🙏
다연님이 리뷰 남겨주신 훅으로 분리하는 부분을 해보았습니다.!
또한 토큰의 유효함을 확인하는 로직은 불필요하여 전부 삭제하였습니다!

+ rebase는 현재 pr에도 적용이 될 수 있나요..? 저도 제 커밋 기록이 너무 헷갈리네요..ㅜ

## 관련 이슈
closes #40

Da-Yeon-Kim committed:
remove: CRA로 생긴 파일 정리

- public 폴더 내 파일 (logo, favicon, manifest.json)
- 기본 CSS 파일 (index.css, App.css)
- 성능 측정 도구 (reportWebVitals.js)
- 테스트 파일 (App.test.js)

 #18

Da-Yeon-Kim committed:
[REMOVE] CRA로 생긴 파일 정리

geongyu09 committed:
docs: pr 템플릿 내용 변경 (#30)

pr 템플릿을 약간 수정하였습니다

HyeonChae1104 committed:
feat(onboarding): 온보딩 화면에 들어갈 이미지와 아이콘 추가

- 온보딩 화면에서 서비스의 기능 설명을 위한 사진을 추가하였습니다.
- 카카오 로그인 버튼을 위한 아이콘을 추가하였습니다.

#31

HyeonChae1104 committed:
feat(onboarding): 온보딩 화면을 구현

- 온보딩 과정에 사용할 이미지를 추가
- 제목, 부제목, 카카오 로그인 버튼을 추가
- 기능 설명을 위해 화살표를 통한 네비게이션 기능을 구현했습니다.

#31

HyeonChae1104 committed:
feat(router): 온보딩 화면을 위한 라우트 설정 추가

- 온보딩 화면으로 이동할 수 있는 라우트를 추가했습니다.

#31

HyeonChae1104 committed:
chore(onboarding): 폰트 변경

폰트를 Pretendard로 변경하였습니다.

#31

Da-Yeon-Kim committed:
feat(Create): 입력 폼 UI 구축

- 모임 이름, 날짜 범위, 예정 소요 시간, 시간 범위 입력 폼 생성
- react-datepicker를 이용하여 캘린더 모달 생성
- react-hook-form을 이용하여 폼 데이터 context로 관리

 #27

Da-Yeon-Kim committed:
chore: 의존성 설치

- 영문은 모르겠지만, react-datepicker, react-hook-form 설치가 안 되어
  있어서 보완

  #27

HyeonChae1104 committed:
feat(onboarding): 작은 화면에서 적용되는 반응형 디자인 추가

- 768px 이하의 화면에서 이미지가 위로 올라가도록 @media 쿼리 추가
- 화면이 작을 때 레이아웃이 수직으로 정렬되도록 수정하였습니다.

#31

HyeonChae1104 committed:
[FEAT] 온보딩 페이지 구현

온보딩 화면을 구현하였습니다. 

- 온보딩 화면은 우리 서비스의 주기능 4가지를 설명하는 역할을 하도록 네비게이션을 구현하였습니다. 
- 화면에 들어갈 이미지와 '카카오 로그인' 버튼을 추가하였습니다.
- 작은 화면에서 적용되는 반응형 디자인을 추가하였습니다. 

#31

Da-Yeon-Kim committed:
design(DateRange): 캘린더 모달 UI 변경

- styled로 덮어 쓰기엔 너무 길어서 별도 css 파일로 관리
- input 폼의 배치 수정
- 캘린더 모달의 색상 수정

 #27

Da-Yeon-Kim committed:
feat(Create): 페이지 UI 요소 배치

 #27

Da-Yeon-Kim committed:
design(Form): 폰트 단위 변경 및 UI 변경

- margin-bottom 대신 gap으로 관리 (그냥 취향)
- 폰트 단위 px 말고 rem으로 관리

 #27

Da-Yeon-Kim committed:
[FEAT] 생성 페이지 폼 UI

- create 페이지 요소 배치 (반응형, 화면이 작아지면 map은 아래로)
- 폼 UI 구축
- react-hook-form, react-datepicker 이용
- 캘린더 모달은 스타일링이 길어져 css 파일을 따로 분리해 관리

geongyu09 committed:
[FEAT] 캘린더 공통 컴포넌트 추가 (#38)

* chore(package.json): 캘린더 구현을 위한 full-calendar 의존성 설치

#35

* feat(useGetMyEvent): 홈 화면에서 사용하는 내 캘린더 가져오기 로직 추가

#35

* feat(calendar.mock.ts): 캘린더 가져오기 목서버 로직 추가

* feat(package.json): full-calendar 의존성 추가

#35

* fix: 올바르게 axios타입이 지정되지 않았던 버그 수정

기존에 resopnse에 존재하던 탬플릿 타입이 기록이 안되어 ts타입과 실제 타입이 맞지 않던 문제 발생

instance에서 이를 해결하기 위하여 data만 사용하도록 수정

#35

* feat(calendar): 반복되어 사용될, 외부 라이브러리 의존성을 가지는 캘린더 UI 컴포넌트 추가

외부 라이브러리에 의존하므로 이를 Service폴더로 뺌
구조의 경우 아래와 동일하다.
- components: 컴포넌트들이 담긴다
- style : css 파일들이 담긴다.
- types : 타입들이 담긴다
- utils : 유틸 파일들이 담긴다
- index.ts : 모듈화 되어있는 파일들을 재모듈화 한다.

현재 utils 내용의 경우 추후 추가하도록 한다.

* feat(Home/index.tsx): 홈 화면에 캘린더 추가

#35

* feat(Calendar): 요구사항의 변경에 따라 불필요한 코드 제거

수정, 삭제가 제거됨에 따라서 불필요한 로직 제거

* refactor(Calendar/utils): 사용하지 않는 유틸 코드 제거

#35

* design(Calendar): 캘린더 디자인 추가

- 자주 사용되는 서비스의 primary 색 변수를 추가하였습니다.
  - 색 변수는 css로 정의되어있으며, 이는 styles/variants/index.ts에서 js로 변환 후 가져오는 구조를 가집니다.
- 캘린더의 기본 스타일을 수정하였습니다.

#35

* refactor: event 오타 수정

* fix: npm 버전이 달라서 발생한 package-lock.json 파일 제거

추후에 이슈 파서 버전을 맞추도록 하겠습니다

#35

* feat: package-lock 파일 생성

geongyu09 committed:
[DOCS] 5주차 리뷰 readme 작성 (#46)

HyeonChae1104 committed:
feat: 카카오로그인버튼 공통 컴포넌트 추가

카카오로그인 버튼은 온보딩 화면과 로그인 로직에서 재사용하기 위해 공통 컴포넌트로 분리하였습니다.

#40

HyeonChae1104 committed:
feat:(Login): 백엔드로 리다이렉트하는 카카오 로그인 기능 추가

카카오 로그인 버튼을 통해 백엔드로 리다이렉트하는 기능을 추가하였습니다.(GET 요청)

#40

HyeonChae1104 committed:
feat(success-page): 액세스 토큰 유효성 검증 및 홈화면 리다이렉트 로직 추가

액세스 토큰을 확인하여 유효하면 홈으로 이동, 그렇지 않으면 실패 페이지로 리다이렉트합니다.

#40

HyeonChae1104 committed:
feat(failure-page): 리프레시 토큰을 통한 재로그인 기능 및 카카오 로그인 버튼 스타일 재사용 추가

에러 메시지를 표시하고, "다시 로그인하기" 버튼을 눌러 리프레시 토큰을 사용해 다시 액세스 토큰을 받아 성공 페이지로 이동합니다.

#40

HyeonChae1104 committed:
feat(button): children 속성 추가 및 카카오 로그인 버튼 컴포넌트 개선

- children 속성 추가로 버튼 안에 동적으로 텍스트 삽입 가능
- 버튼 디자인 재사용성 강화
- 스타일 컴포넌트 유지하며 클릭 이벤트 핸들링 기능 보완하였습니다.

#40

HyeonChae1104 committed:
feat(routes): 성공 및 실패 페이지 경로 추가

성공/실패 페이지 경로 추가 및 경로 설정을 하였습니다.

#40

geongyu09 commented:
커밋이 다른 사람들 것 까지 쌓여서 보기가 힘드네요..
어디 파일을 집중적으로 보면 될까요?

HyeonChae1104 commented:
- [ src/pages/Login/index.tsx](https://github.com/kakao-tech-campus-2nd-step3/Team16_FE/blob/00d5330fc7ee87dbf18201557342fa1759647089/src/pages/Login/index.tsx)
- [ src/pages/Login/SuccessPage.tsx](https://github.com/kakao-tech-campus-2nd-step3/Team16_FE/blob/00d5330fc7ee87dbf18201557342fa1759647089/src/pages/Login/SuccessPage.tsx)
- [ src/pages/Login/FailurePage.tsx](https://github.com/kakao-tech-campus-2nd-step3/Team16_FE/blob/00d5330fc7ee87dbf18201557342fa1759647089/src/pages/Login/FailurePage.tsx)
- [ routes/index.tsx](https://github.com/kakao-tech-campus-2nd-step3/Team16_FE/blob/00d5330fc7ee87dbf18201557342fa1759647089/src/routes/index.tsx)
- [routes/path.ts](https://github.com/kakao-tech-campus-2nd-step3/Team16_FE/blob/00d5330fc7ee87dbf18201557342fa1759647089/src/routes/path.ts)
- [src/components/common/Button/kakaoLogin.tsx
](https://github.com/kakao-tech-campus-2nd-step3/Team16_FE/blob/00d5330fc7ee87dbf18201557342fa1759647089/src/components/common/Button/kakaoLogin.tsx)

보기 힘드셨겠네용 죄송합니다..ㅜ
이 파일들을 수정 및 추가하였습니다! 이 파일들 위주로 확인 부탁드립니다!

Da-Yeon-Kim commented on src/routes/index.tsx:
```diff
@@ -1,53 +1,70 @@
 import { HomePage } from '@/pages/Home';
 import { JoinPage } from '@/pages/Join';
 import { LeaderPage } from '@/pages/Leader';
 import { LoginPage } from '@/pages/Login';
```
`LoginPage`가 현재 default export를 하고 있는데 `{ LoginPage }`로 import 해와서 오류가 발생하네요.
개인적으로는 다른 페이지들이 현재 named export 방식이기 때문에,
`LoginPage`, `FailurePage`, `SuccessPage` 모두 그에 맞추어 바꾸어주는 게 좋을 것 같습니다~

Da-Yeon-Kim commented on src/routes/index.tsx:
```diff
@@ -1,53 +1,70 @@
   return <RouterProvider router={router} />;
 };
+
+export const Routes = RoutesComponent;
```
혹시 라우팅 처리를 컴포넌트로 바꾸어서 처리하신 이유가 있을까요?

HyeonChae1104 replied:
다시 수정해두었습니다.. 하하

Da-Yeon-Kim commented on src/pages/Login/FailurePage.tsx:
```diff
@@ -0,0 +1,34 @@
+
+  const retryLogin = async () => {
+    try {
+      const response = await axios.post('/api/refresh-token', null, {
```
이미 만들어진 `fetchInstance`를 사용해도 될 것 같은데, 혹시 사용하지 않으신 이유가 있을까요?

그리고 토큰과 관련된 작업은 별도의 훅으로 분리해서 사용하는 편이 폴더 구조에 더 어울릴 것 같습니다~

Da-Yeon-Kim commented on src/pages/Login/SuccessPage.tsx:
```diff
@@ -0,0 +1,31 @@
+        const accessToken = localStorage.getItem('accessToken');
+        if (!accessToken) throw new Error('토큰 없음');
+
+        await axios.get('/api/verify-token', {
```
이 코드도 `fetchWithToken`을 써도 괜찮을 것 같은데, 혹시 뭔가 다른가요? 잘 몰라서 물어보는 거니 알려주시면 감사히 배우겠습니다!

+) 이것도 별도로 훅으로 관리하는 게 좋아 보여요😊

Da-Yeon-Kim commented on src/pages/Onboarding/index.tsx:
```diff
@@ -0,0 +1,181 @@
+  align-items: center;
+  justify-content: space-between;
+  
+  @media (max-width: 768px) {
```
기존에 만들어둔 `breakpoints`를 사용하면 더 좋을 것 같아요~

Da-Yeon-Kim commented on src/pages/Login/index.tsx:
```diff
@@ -1,3 +1,17 @@
+
+const LoginPage: React.FC = () => {
+  const handleKakaoLogin = () => {
+    window.location.href = 'http://localhost:8080/api/auth/login';
```
`http://localhost:8080`이 필요하시면 .env 파일에서 `REACT_APP_API_URL`을 사용하시는 게 맞아 보여요!

그런데 제가 잘 몰라서 그러는데 `api/auth/login`은 API 통신 주소 아닌가요...? 혹시 이 부분이 어떤 동작인지 설명해주실 수 있나요...? 로그인 로직을 몰라서 피드백이 어렵네요😅

Da-Yeon-Kim reviewed (commented):
로그인 로직이 저는 아직도 어렵네요😅 고생하셨어요!
제가 이해를 잘 못해서 질문이 좀 많은데, 양해 부탁드립니다🙏

혹시 최신 사항 `fetch` 해온 다음에 `rebase` 하셨나요? 커밋 기록이 많아서 보기가 조금 어렵긴 하네요! 만약 안 하셨다면 한 번 시도해보시는 걸 추천 드립니다👍

page 코드에 모두 작성하시기보단 기왕이면 컴포넌트를 적절히 분배하고 파일 구조에 맞추어 훅들을 따로 분리해 정리해주시면 더 좋을 것 같아요~

HyeonChae1104 committed:
refactor(SuccessPage): SuccessPage의 로직을 훅으로 분리

- getCookieValue를 통해 accessToken을 쿠키에서 가져옵니다.
-accessToken 확인 후 localStorage에 저장하고, 보안을 위해 그 값을 바로 삭제합니다.
- accessToken이 존재하면 홈 페이지(/)로 이동하고, 존재하지 않으면 실패 페이지(/failure)로 이동합니다.

#40

HyeonChae1104 committed:
refactor(FailurePage): FailurePage의 로직을 훅으로 분리

- named export 방식으로 수정
- 백엔드에서 로그인 실패할 경우 redirect 되는 페이지로 '카카오로그인버튼'을 누르면 다시 백엔드의 로그인 로직으로 넘어갈 수 있도록 합니다.
- 에러 상황을 처리했던 로직은 프론트에서 불필요하여 전부 삭제하였습니다.

#40

HyeonChae1104 committed:
feat(authService): 401 응답 시 토큰 재발급을 위한 useFetchWithToken 추가

- accessToken 만료 시 토큰 재발급을 처리하는 useFetchWithToken 훅 생성
- Axios 인터셉터를 통해 토큰 재발급 후 요청을 자동 재시도하도록 구현
- 토큰 재발급 실패 시 로그인 페이지로 리다이렉트

#40

HyeonChae1104 committed:
fix(login):

-  login 로직 변경에 의한 import 경로 추가
- named export로 통일

#40

HyeonChae1104 committed:
refactor(useSuccessLogin): 로그인 성공 시 쿠키 처리 로직 추가

- 로그인 성공 후 토큰 처리 기능 구체화하였습니다.

geongyu09 commented:
저번에 기능 구현 이후에 알려주신다고 하셨어서 리뷰를 안 달고 있었습니다!
달아도 될 때쯤에 다시 말씀해 주시면 감사할 것 같습니다
